From 8d42db6ea09007f98c107694a8933a14a3aab5e8 Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Thu, 22 Mar 2018 17:08:10 -0500
Subject: [PATCH 3535/5242] MLK-17703-2: drm: change HDR metadata infoframe
 structure

commit  96f958b789b0bccbc41062d2a1c1b2c80d2ab7b7 from
https://source.codeaurora.org/external/imx/linux-imx.git

According to ANSI-CTA-861-G specification:
 * EOTF is 8 bit, not 16;
 * metadata type is 8 bit, not 16;
 * There's no "Minimum Content Light Level"

This patch will change the HDR metadata structures to reflect that. Also, this
will fix problems seen on some TVs that were rejecting HDR metadata because
it's size was too big (more than 26 bytes).

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
CC: Sandor Yu <sandor.yu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/drm_edid.c  |    6 +-----
 include/linux/hdmi.h        |    5 ++---
 include/uapi/drm/drm_mode.h |    5 ++---
 3 files changed, 5 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c
index 7f19a47..d3aff75 100644
--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -3888,11 +3888,7 @@ static uint16_t hdr_metadata_type(const u8 *edid_ext)
 	info->hdr_panel_metadata.eotf = eotf_supported(db);
 	info->hdr_panel_metadata.type = hdr_metadata_type(db);
 
-	if (len == 6) {
-		info->hdr_panel_metadata.max_cll = db[4];
-		info->hdr_panel_metadata.max_fall = db[5];
-		info->hdr_panel_metadata.min_cll = db[6];
-	} else if (len == 5) {
+	if (len == 5) {
 		info->hdr_panel_metadata.max_cll = db[4];
 		info->hdr_panel_metadata.max_fall = db[5];
 	} else if (len == 4) {
diff --git a/include/linux/hdmi.h b/include/linux/hdmi.h
index 62b9bde..7ddc93e 100644
--- a/include/linux/hdmi.h
+++ b/include/linux/hdmi.h
@@ -165,8 +165,8 @@ struct hdmi_drm_infoframe {
 	enum hdmi_infoframe_type type;
 	unsigned char version;
 	unsigned char length;
-	uint16_t eotf;
-	uint16_t metadata_type;
+	uint8_t eotf;
+	uint8_t metadata_type;
 	uint16_t display_primaries_x[3];
 	uint16_t display_primaries_y[3];
 	uint16_t white_point_x;
@@ -175,7 +175,6 @@ struct hdmi_drm_infoframe {
 	uint16_t min_mastering_display_luminance;
 	uint16_t max_fall;
 	uint16_t max_cll;
-	uint16_t min_cll;
 };
 
 int hdmi_avi_infoframe_init(struct hdmi_avi_infoframe *frame);
diff --git a/include/uapi/drm/drm_mode.h b/include/uapi/drm/drm_mode.h
index 37392bb..1cd4a9c 100644
--- a/include/uapi/drm/drm_mode.h
+++ b/include/uapi/drm/drm_mode.h
@@ -630,8 +630,8 @@ enum supported_eotf_type {
 
 /* HDR Metadata */
 struct hdr_static_metadata {
-	uint16_t eotf;
-	uint16_t type;
+	uint8_t eotf;
+	uint8_t type;
 	uint16_t display_primaries_x[3];
 	uint16_t display_primaries_y[3];
 	uint16_t white_point_x;
@@ -640,7 +640,6 @@ struct hdr_static_metadata {
 	uint16_t min_mastering_display_luminance;
 	uint16_t max_fall;
 	uint16_t max_cll;
-	uint16_t min_cll;
 };
 
 #define DRM_MODE_PAGE_FLIP_EVENT 0x01
-- 
1.7.9.5

